game/src/sandbox/dashboards/commuter.rs

```diff
@@ -284,52 +231,201 @@ impl State for CommuterPatterns {
-
-                let new_scale = ColorLegend::gradient(
-                    ctx,
-                    &app.cs.good_to_bad_red,
-                    vec![
-                        "0".to_string(),
-                        format!("{} trips", prettyprint_usize(max_cnt)),
-                    ],
-                )
-                .named("scale");
-                self.composite.replace(ctx, "scale", new_scale);
             } else {
-                self.composite.replace(
-                    ctx,
-                    "current",
-                    "None selected".draw_text(ctx).named("current"),
-                );
+                // Hovering over block
+                match self.current_block.0 {
+                    BlockSelection::Locked { base, .. } => {
+                        if base == b.id {
+                            BlockSelection::Locked {
+                                base,
+                                compare_to: None,
+                            }
+                        } else {
+                            BlockSelection::Locked {
+                                base,
+                                compare_to: Some(b.id),
+                            }
+                        }
+                    }
+                    BlockSelection::Unlocked(_) => BlockSelection::Unlocked(b.id),
+                    BlockSelection::NothingSelected => BlockSelection::Unlocked(b.id),
+                }
+            }
+        } else {
+            // cursor not over any block
+            match self.current_block.0 {
+                BlockSelection::NothingSelected | BlockSelection::Unlocked(_) => {
+                    BlockSelection::NothingSelected
+                }
+                BlockSelection::Locked { base, .. } => BlockSelection::Locked {
+                    base,
+                    compare_to: None,
+                },
+            }
+        };
+
+        let mut filter = Filter {
+            from_block: self.composite.is_checked("from / to this block"),
+            include_borders: self.composite.is_checked("include borders"),
+            depart_from: app
+                .primary
+                .sim
+                .get_end_of_day()
+                .percent_of(self.composite.area_slider("depart from").get_percent()),
+            depart_until: app
+                .primary
+                .sim
+                .get_end_of_day()
+                .percent_of(self.composite.area_slider("depart until").get_percent()),
+            modes: BTreeSet::new(),
+        };
+        for m in TripMode::all() {
+            if self.composite.is_checked(m.ongoing_verb()) {
+                filter.modes.insert(m);
             }
         }
 
+        if filter != self.filter || block_selection != self.current_block.0 {
+            self.filter = filter;
+            let (drawable, per_block_counts) = self.build_block_drawable(block_selection, ctx, app);
```

Comment:
I'm not completely happy with how this ended up - build_block_drawable is an immutable method which takes in some state and returns the drawable.

I put redraw_composite as a separate method for two reasons... it does do some mutation and I thought it made sense to group all the code together which deals with drawing that part of the screen. 

The kind of awkward thing is that it uses some of the per-block state. Rather than re-compute it in that method, I'm re-using the values computed from build_block_drawable.

If you find this offensive, I'm happy to either:
1. recompute the per_block_counts state in redraw_composite
2. or mash build_block_drawable and redraw_compositor into one (mutable) method
3. or maybe you have a better suggestion...

Reply:
The way you've done it seems fine to me. The other alternative is to recreate the entire Composite from scratch, to whittle down some of the replace calls. But I don't think that winds up looking simpler, because then there's some effort to turn `Filter` into the correct initial state.